src/core/namespace.c

```diff
@@ -287,12 +288,31 @@ static int apply_mount(
 
                 /* First, get rid of everything that is below if there
                  * is anything... Then, overmount it with an
-                 * inaccessible directory. */
+                 * inaccessible path. */
                 umount_recursive(m->path, 0);
 
-                what = "/run/systemd/inaccessible";
-                break;
+                /* mount() always follow symlinks,
+                 * so we follow symlinks during the stat() as well */
+                r = stat(m->path, &target);
+                if (r != 0)
+                        return -errno;
 
+                if (S_ISREG(target.st_mode))
+                        what = SD_INACCESSIBLE_PATH_REG;
+                else if (S_ISDIR(target.st_mode))
+                        what = SD_INACCESSIBLE_PATH_DIR;
+                else if (S_ISCHR(target.st_mode))
+                        what = SD_INACCESSIBLE_PATH_CHR;
+                else if (S_ISBLK(target.st_mode))
+                        what = SD_INACCESSIBLE_PATH_BLK;
+                else if (S_ISFIFO(target.st_mode))
+                        what = SD_INACCESSIBLE_PATH_FIFO;
+                else if (S_ISSOCK(target.st_mode))
+                        what = SD_INACCESSIBLE_PATH_SOCK;
+                else {
+                        // TODO warning, error, return without mount?
```

Comment:
I think failing with an error is fine in this case. I mean, this can only happen or symlinks, after all, and I think it would to treat that as configuration error...